Enabled multivariate blocks

grishabh147 opened the pull request:
I have made the following changes/additions in the code:
- Enabled multivariate blocks
- Enable multiple constraints in a block
- Used hypergraph abstraction to store the BlockNLPModel structure
- Added a test problem based on MPC for a chemical reactor
- Added docstrings

I am having some troubles with IPOPT on my computer and haven't been able to run full tests on this code. However, I have tested individual functions and they seem to be generating correct output. I will post an update whenever I am able to resolve the IPOPT issue.

My plan now is to also implement ADMM before we merge this to the main branch.

grishabh147 committed:
Create Full_Space.jl

grishabh147 committed:
First commit

grishabh147 committed:
Delete Full_Space.jl

grishabh147 committed:
Initial commit

grishabh147 committed:
Merge branch 'rg/develop' of https://github.com/exanauts/BlockNLPModels.jl into rg/develop

grishabh147 committed:
remove comments

grishabh147 committed:
Update Full_Space.jl

grishabh147 committed:
add TOML files

grishabh147 committed:
Update Project.toml

grishabh147 committed:
Create module

grishabh147 committed:
Updated files after 1st PR

grishabh147 committed:
Addressed all review comments

grishabh147 committed:
Updated abstract type name

grishabh147 committed:
Enable multivariable blocks

grishabh147 committed:
made linking_constraints modular

grishabh147 committed:
Fixed full_space.jl's derivatives

grishabh147 committed:
added docstrings

grishabh147 committed:
updated docstrings

grishabh147 committed:
added docstrings

grishabh147 committed:
Added docstrings

grishabh147 committed:
added docstrings

grishabh147 committed:
Add docstrings

grishabh147 committed:
dual decomposition algorithm for multivariate blocks

grishabh147 committed:
Update dual_decomposition.jl

grishabh147 committed:
Make compatible with new functions

frapac commented on docs/make.jl:
```diff
@@ -0,0 +1,3 @@
+push!(LOAD_PATH,"../src/")
```
We should remove this harcoded path

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,213 @@
+using LinearAlgebra
+using SparseArrays
+
+export AbstractBlockNLPModel
```
I would be careful here, and limit the number of `export` (ideally we should have 0 functions exported). A common rule is to not export the abstract types.

grishabh147 replied:
I think exporting ``AbstractBlockNLPModel`` here is unavoidable, because, eventually ``BlockNLPAlgorithms`` will be written with an understanding that a model being input to that package is an ``AbstractBlockNLPModel``. Won't we need this abstract type at that time?

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,213 @@
+include("utils.jl")
+
+"""
+  BlockNLP_Counters()
```
To follow the [Julia style guide](https://docs.julialang.org/en/v1/manual/style-guide/), I would rename this function `BlockNLPCounters` (we should avoid underscore when using CamelCase names)

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,213 @@
+  BlockNLP_Counters()
+Keeps a count of blocks, linking constraints, variables, and (block) constraints attached to the model.
+"""
+mutable struct BlockNLP_Counters
```
Careful with the indentation here! It is usually preferred to use 4 spaces for indentations

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,213 @@
+    AbstractLinearLinkConstraint
+Abstract supertype for the definition of linear linking constraints
+"""
+abstract type AbstractLinearLinkConstraint end
```
Do we need to define this abstract type? If we have only one struct inheriting from this abstract type, it would not make sense.

grishabh147 replied:
You're right, there is no use of this abstract type right now. But later, once we also have nonlinear linking constraints, I thought it would be useful to have this type definition to differentiate between the two types.

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,213 @@
+end
+
+"""
+    add_links(block_nlp::AbstractBlockNLPModel, 
```
I would reformat as:
```julia
    add_links(
        block_nlp::AbstractBlockNLPModel, 
        n_constraints::Int,   
        links::Union{Dict{Int, Array{Float64, 2}}, Dict{Int, Vector{Float64}}}, 
        constants::Union{AbstractVector, Float64},
    )
```

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,213 @@
+function add_block(block_nlp::AbstractBlockNLPModel, nlp::AbstractNLPModel)
+  block_nlp.problem_size.block_counter += 1
+
+  temp = deepcopy(block_nlp.problem_size.var_counter)+1
```
Do we really need this machinery?

grishabh147 replied:
I am using this to assign a unique index to all block variables. Having a unique index assigned to every variable was very useful while writing ``fullspace.jl``, ``dualized_block.jl``. I am not sure if there is a smarter way that can avoid this. I am open to suggestions.

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,213 @@
+- `links::Union{Dict{Int, Array{Float64, 2}}, Dict{Int, Vector{Float64}}}`: coefficients of the block matrices that link different blocks,
+- `constants::Union{AbstractVector, Float64}`: RHS vector
+"""
+# 
```
I would remove this line with the `#`

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,213 @@
+  constants::Union{AbstractVector, Float64})
+  
+  linking_blocks = Vector{SparseMatrixCSC{Float64, Int}}(undef, block_nlp.problem_size.block_counter)
+  RHS_vector = Vector{Float64}(undef, n_constraints)
```
rename as `rhs_vector`

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,213 @@
+  # Initialize linking_blocks with empty matrices
+  for i in 1:block_nlp.problem_size.block_counter
+    linking_blocks[i] = spzeros(n_constraints, block_nlp.blocks[i].meta.nvar)
+  end
```
In that case, we can instantiate `linking_blocks` directly as:

```julia
linking_blocks = [spzeros(n_constraints, block_nlp.blocks[i].meta.nvar) for i in 1:block_nlp.problem_size.block_counter]
```
(in general it is better to avoid dealing with `undef` array)

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,213 @@
+  link_con_idx = Vector{Int}()
+
+  # Parse the constraints one-by-one to collect this information
+  temp = block_nlp.problem_size.link_counter+1
```
I would prefer we avoid naming a variable `temp`

frapac commented on src/dualized_block.jl:
```diff
@@ -0,0 +1,78 @@
+"""
+    DualizedNLPblockModel{T, S} <: AbstractNLPModel{T, S}   
```
I would rename as `DualizedNLPBlockModel` to be consistent in the naming

frapac commented on src/dualized_block.jl:
```diff
@@ -0,0 +1,78 @@
+
+A data type to store dualized subproblems. 
+"""
+mutable struct DualizedNLPblockModel{T, S} <: AbstractNLPModel{T, S}
```
Yes, that's exactly the structure we need!

frapac commented on src/dualized_block.jl:
```diff
@@ -0,0 +1,78 @@
+
+function NLPModels.obj(nlp::DualizedNLPblockModel, x::AbstractVector)
+  n = nlp.problem_block.meta.nvar
+  d(x) = dot(nlp.λ, nlp.A, x)
```
Why building a closure here? We can directly return:
```julia
    return obj(nlp.problem_block, x) + dot(nlp.λ, nlp.A, x)
```

frapac commented on src/dualized_block.jl:
```diff
@@ -0,0 +1,78 @@
+
+function NLPModels.grad!(nlp::DualizedNLPblockModel, x::AbstractVector, g::AbstractVector)
+  grad!(nlp.problem_block, x, g)
+  g .+= nlp.A'*nlp.λ
```
I would replace this line by using `mul!` directly:
```julia
mul!(g, nlp.A', nlp.λ, 1, 1)
```

grishabh147 replied:
Won't this overwrite the values stored in `g`? Whereas what I want to do here is to add `nlp.A'*nlp.λ` to the existing elements of `g`.

frapac replied:
The `mul!` syntax is a bit misleading in Julia. `mul!` with three arguments does overwrite, but `mul!` with 5 arguments modifies the values inplace. See the documentation.

frapac commented on src/full_space.jl:
```diff
@@ -0,0 +1,224 @@
+
+- `m::AbstractBlockNLPModel`: name of the BlockNLPModel
+"""
+function FullSpaceModel(::Type{T}, m::AbstractBlockNLPModel) where {T}
```
Why do we need `Type{T}` in the signature? Are we planning to work in multiprecision?

grishabh147 replied:
We don't need it. I wrote a different version earlier where I needed to keep `Type{T}` there, but later changed it. I have removed it now. Thanks for pointing this out.

frapac commented on src/full_space.jl:
```diff
@@ -0,0 +1,224 @@
+        minimize = true,
+        name = "full_space",
+    )
+return FullSpaceModel(meta, Counters(), m)
```
indentation issue

frapac commented on src/full_space.jl:
```diff
@@ -0,0 +1,224 @@
+    increment!(nlp, :neval_grad)
+    for i in 1:nb
+        gx[nlp.blocknlp.blocks[i].var_idx] = 
+        grad(nlp.blocknlp.blocks[i].problem_block, x[nlp.blocknlp.blocks[i].var_idx])
```
Can we modify `gx` inplace here by using `grad!`? In general, it's better to avoid any new allocation of a vector

grishabh147 replied:
We can, I have updated the code. One question: Isn't `gx` already allocated here? and I am just assigning new values to its elements.

grishabh147 replied:
I tried doing that, but it somehow doesn't work. Also, unless I am misunderstanding your comment, I don't think I am allocating any new memory here, because `gx` is already initialized as a `nlp.meta.nvar` sized vector.

grishabh147 replied:
Okay, this is done now using `@view`.

frapac commented on src/full_space.jl:
```diff
@@ -0,0 +1,224 @@
+    nb = nlp.blocknlp.problem_size.block_counter
+    @lencheck nlp.meta.nnzh rows cols
+    H = spzeros(nlp.meta.nvar, nlp.meta.nvar)
+    for i in 1:nb
```
This code is very difficult to read

grishabh147 replied:
I have made some changes. Hopefully, it will read better now.

frapac replied:
it reads better!

frapac commented on src/full_space.jl:
```diff
@@ -0,0 +1,224 @@
+
+    H = spzeros(nlp.meta.nvar, nlp.meta.nvar)
+    for i in 1:nb
+    H[nlp.blocknlp.blocks[i].var_idx, nlp.blocknlp.blocks[i].var_idx] = 
```
This code is almost exactly the same as in the previous function. Can we define an intermediate function to avoid duplicating code?

grishabh147 replied:
Done.

frapac commented on src/full_space.jl:
```diff
@@ -0,0 +1,224 @@
+    idx = 0
+    for i in 1:nb
+        if nlp.blocknlp.blocks[i].meta.ncon > 0
+            temp = deepcopy(idx)+1
```
`idx`is suppose to be an `Int`. Why do we need to deepcopy it here?

grishabh147 replied:
We don't. I copied this line from somewhere else, and left the `deepcopy()` as it is.

frapac commented on src/utils.jl:
```diff
@@ -0,0 +1,83 @@
+    A = spzeros(m.problem_size.link_counter, m.problem_size.var_counter)
+    for i in 1:nb
+        for j in 1:length(m.linking_constraints)
+            # if m.linking_constraints[j] isa AbstractLinearLinkConstraint
```
remove dead comment

frapac commented on src/utils.jl:
```diff
@@ -0,0 +1,83 @@
+"""
+  get_blockmatrix(
```
How about renaming this function `get_linking_matrix` ?

frapac commented on src/utils.jl:
```diff
@@ -0,0 +1,83 @@
+    m::AbstractBlockNLPModel 
+  )
+
+Returns a vector of block matrices `[A_1, \\ldots, A_B]``
```
Slight typo: 
```
``[A_1, \\ldots, A_B]``
```

frapac commented on src/utils.jl:
```diff
@@ -0,0 +1,83 @@
+end
+
+"""
+    get_RHSvector(
```
I would rename this function as `get_rhs_vector` to respect the naming convention

frapac commented on src/utils.jl:
```diff
@@ -0,0 +1,83 @@
+end
+
+"""
+    n_tot_con(
```
The name is not very clear. How about `n_constraints`?

frapac reviewed (changes requested):
I think we are heading in the right direction. I mostly have minor comments. The new tests are deeply appreciated. 

The PR can be merged once the comments have been addressed. Also, let me know if you need help to fix the conflict issue.

grishabh147 committed:
Implemented ADMM

grishabh147 committed:
Added tests, addressed PR comments

grishabh147 commented:
The latest commit brings the following updates:
- Added tests that are exclusive for ``BlockNLPModels.jl``
- Made the changes requested by @frapac
- Removed ``dual_decomposition.jl`` and ``admm.jl`` from this repo. I will create a new repo for ``BlockNLPAlgorithms.jl`` and put these files there.

As you will see in ``simple_block_model.jl``, even ``MadNLP.jl`` is not able to solve the full-space model created by ``FullSpaceModel()``. I think there is some bug in my code that I am not able to figure out. Looking at the stacktrace makes me think that the issue is related to the size of sparse Hessian matrix that my code is generating. Can someone please help me debug this issue?

grishabh147 committed:
Fixed Project.toml

grishabh147 committed:
Slightly modified hessian, jacobian code

grishabh147 committed:
debugged, finally works with IPOPT

grishabh147 committed:
passes all tests

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,214 @@
+    AbstractNLLinkConstraint
+Abstract supertype for the definition of nonlinear linking constraints
+"""
+abstract type AbstractNLLinkConstraint end
```
I would suggest to rename this as `AbstractNonlinearLinkConstraint`

frapac commented on src/BlockNLPModels.jl:
```diff
@@ -0,0 +1,214 @@
+    rhs_vector::Vector{Float64}
+end
+
+mutable struct NLLinkConstraint <: AbstractNLLinkConstraint
```
Are we using this at the moment? If not, I would suggest to remove it for now.

frapac commented on src/augmented_block.jl:
```diff
@@ -0,0 +1,185 @@
+      ones(nlp.meta.nnzh), 
+      nlp.meta.nvar, 
+      nlp.meta.nvar
+    ) + ρ.*abs.(A[:, nlp.var_idx]'*A[:, nlp.var_idx])
```
The formatting is better now!

frapac commented on src/augmented_block.jl:
```diff
@@ -0,0 +1,185 @@
+    obj_weight::Union{Number, Nothing} = nothing, 
+    vals::Union{AbstractVector, Nothing} = nothing)
+
+    H = spzeros(m.meta.nvar, m.meta.nvar)
```
I don't think we need this line as we are redefining `H` two lines after

frapac commented on src/augmented_block.jl:
```diff
@@ -0,0 +1,185 @@
+      nlp.meta.nvar
+    ) + m.ρ.*abs.(m.A[:, m.subproblem.var_idx]'*m.A[:, m.subproblem.var_idx])
+    rows, cols, temp_vals = findnz(H)
+    if x === nothing
```
You can use directly the function `isnothing` here

frapac reviewed (approved):
The PR looks good to me. It remains only a few minor issues to address (like fixing the conflicts). I suggest to do that together during our meeting today.

grishabh147 committed:
Updated with minor fixes

grishabh147 committed:
Addressed new PR comments

frapac committed:
Merge remote-tracking branch 'origin/main' into multivariate_blocks

frapac committed:
fix formatting
